lib/signaling/v3/room.js

```diff
@@ -0,0 +1,158 @@
+    }
+    return switchOffState;
+  }
+
+  /**
+   * @private
+   */
+  _getPendingTrackReceiver(trackSid) {
+    const mid = this._pendingTrackMids.get(trackSid);
+    if (!mid) {
+      return Promise.resolve(null);
+    }
+    this._pendingTrackMids.delete(trackSid);
+    return this._getTrackReceiver(mid, 'mid');
+  }
+
+  /**
+   * @private
+   * @override
+   */
+  _handleSubscriptions() {
+    /* Do nothing since RSP v3 messages will not contain the "subscribed" property. */
+  }
+
+  /**
+   * @private
+   */
+  _initTrackSubscriptionsSignaling() {
+    this._trackSubscriptionsSignaling.on('updated', (subscribed, errors) => {
+      const trackSidsToTrackSignalings = this._getTrackSidsToTrackSignalings();
+      const trackSidsToTrackStates = new Map(Object.entries(subscribed));
+      const trackSidsToErrors = new Map(Object.entries(errors));
+
+      trackSidsToTrackStates.forEach(({ mid, off_reason: switchOffReason = null, state }, sid) => {
+        const trackSignaling = trackSidsToTrackSignalings.get(sid);
+        const trackState = { state, switchOffReason };
+        if (!trackSignaling) {
+          this._pendingSwitchOffStates.set(sid, trackState);
+          this._pendingTrackMids.set(sid, mid);
+          return;
+        }
+        const isSwitchedOff = state === 'OFF';
+        if (isSwitchedOff || (trackSignaling.trackTransceiver && trackSignaling.trackTransceiver.mid !== mid)) {
+          // NOTE(mmalavalli): If a RemoteTrackPublicationV3's MID changes, then we need to unsubscribe
+          // from the RemoteTrack before subscribing to it again with the MediaTrackReceiver associated with the new
+          // MID. If a RemoteTrackPublicationV3's RemoteTrack is switched off, then we should still be subscribed
+          // to it, even though it no longer has an MID associated with it.
+          trackSignaling.setTrackTransceiver(null, isSwitchedOff);
+        }
+        if (!isSwitchedOff) {
+          this._getTrackReceiver(mid, 'mid').then(trackReceiver => trackSignaling.setTrackTransceiver(trackReceiver, true));
+        }
+        trackSignaling.setSwitchedOff(isSwitchedOff);
+      });
+
+      trackSidsToErrors.forEach(({ code, message }, sid) => {
+        const trackSignaling = trackSidsToTrackSignalings.get(sid);
+        if (trackSignaling) {
+          trackSignaling.subscribeFailed(createTwilioError(code, message));
+        }
+      });
+
+      trackSidsToTrackSignalings.forEach(trackSignaling => {
+        const { sid } = trackSignaling;
+        if (!trackSidsToTrackStates.has(sid)) {
+          this._pendingSwitchOffStates.delete(sid);
+          this._pendingTrackMids.delete(sid);
+          trackSignaling.setTrackTransceiver(null, false);
+        }
+      });
+    });
+  }
+
+  /**
+   * @private
+   * @override
+   */
+  _updateSubscribed(roomState) {
+    /* Do nothing since RSP v3 messages will not contain the "subscribed" property. */
+  }
```

Comment:
Maybe subclassing will prevent you from adopting future RSP v3 changes due to dependency on RSP v2. Just a thought anyways..

Reply:
@ceaglest Yes, I agree. Most of the code shared between RoomV3 and RoomV2 should actually be in the RoomSignaling base class. However, it is a very complex undertaking to move all that code from RoomV2 to RoomSignaling, so I will do so when it comes time to get rid of RoomV2 altogether.